### PR TITLE
decompose --rootless into $DOCKER_HONOR_XDG and --exec-opt native.*

### LINKED ISSUE #38702: Rootful-Docker-in-Rootless-Docker doesn't work
**Description**

Rootful docker in Rootless docker does not work

**Steps to reproduce the issue:**
1. Start rootless docker
2. Within rootless docker, start `dockerd` as the root(-in-userns).
3. Run `docker run -it --rm busybox` against Rootful-Docker-in-Rootless-Docker

**Describe the results you received:**
```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "process_linux.go:275: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/cpuset/docker: permission denied\"": unknown.
```

**Describe the results you expected:**
It should work

**Additional information you deem important (e.g. issue happens only occasionally):**

Running Rootful-Docker-in-Rootless-Docker daemon with `--experimental --rootless` doesn't make sense here currently, as it tries to use `$XDG_RUNTIME_DIR` and so on.

Probably we should let `--rootless` disable cgroups but ignore `$XDG_RUNTIME_DIR` when `$USER=="root" || $USER==""`.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           18.09.1-rc1
 API version:       1.39
 Go version:        go1.10.5
 Git commit:        bca0068
 Built:             Fri Dec  7 05:28:04 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          dev
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.11.5
  Git commit:       93d994e29c
  Built:
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.0-243-g6b25c1e4
  GitCommit:        6b25c1e45c2b8246dba17de3b1d574f6720ce79f
 runc:
  Version:          1.0.0-rc6+dev
  GitCommit:        bbb17efcb4c0ab986407812a31ba333a7450064c
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: dev
Storage Driver: vfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6b25c1e45c2b8246dba17de3b1d574f6720ce79f
runc version: bbb17efcb4c0ab986407812a31ba333a7450064c
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.15.0-45-generic
Operating System: Ubuntu 18.04.1 LTS (containerized)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.829GiB
Name: e9a9508c8101
ID: UZUL:JHFE:3567:N2FE:YUZL:XKRW:EKQB:I35U:MTDM:7Y3Z:3EO6:DWL5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**- What I did**

Fix #38702 #39009

**- How I did it**

The `--rootless` monolith did not work well for several edge cases:
 * #38702: euid=0, $USER="root" but no access to cgroup ("rootful" Docker in rootless Docker)
 * #39009: euid=0 but $USER="docker" (rootful boot2docker)

 This commit decompose the `--rootless` monolith to the following environment variable and flags:
 * `$DOCKER_HONOR_XDG`: honor `$XDG_RUNTIME_DIR`, `$XDG_DATA_HOME`, and `$XDG_CONFIG_HOME` for detecting the default dirs
 * `--exec-opt native.cgroupdriver=none`: disable cgroups (needs `--experimental` to be specified together)
 * `--exec-opt native.restrict_oom_score_adj=1`: restrict oom_score_adj value (needs `--experimental` to be specified together)

`dockerd` doesn't try to auto-detect these configuration but now `dockerd-rootless.sh` launches `dockerd` with "rootless flags" by default: `--experimental --exec-opt native.cgroupdriver=none --exec-opt native.restrict_oom_score_adj=1 --userland-proxy --userland-proxy-path=$which rootlesskit-docker-proxy)`

**- How to verify it**

```console
$ dockerd-rootless.sh
```

```console
$ docker -H unix://$XDG_RUNTIME_DIR/docker.sock info
...
Cgroup Driver: none
...
```
(Note: `Security Options` no longer contains "rootless")

```
$ docker -H unix://$XDG_RUNTIME_DIR/docker.sock  run ...
```

**- Description for the changelog**

decompose --rootless into $DOCKER_HONOR_XDG and --exec-opt native.*

**- A picture of a cute animal (not mandatory but encouraged)**
:penguin:

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @tianon @tonistiigi @tiborvass @seemethere @andrewhsu

File: contrib/dockerd-rootless.sh [hunk 1]
```diff
@@ -13,6 +10,9 @@
 #
 # See the documentation for the further information.
 
+# The default flags used for running dockerd
+: ${DOCKERD_ROOTLESS_DOCKERD_FLAGS:="--experimental --exec-opt native.cgroupdriver=none --exec-opt native.restrict_oom_score_adj=1 --userland-proxy --userland-proxy-path=$(which rootlesskit-docker-proxy)"}
```
Comment (AkihiroSuda) [inline contrib/dockerd-rootless.sh]: @seemethere @andrewhsu (?)

Please update docker-rootless-extras.tgz to include rootlesskit-docker-proxy before merging this PR.

**Comment by cpuguy83** [comment]
I applaud the desire to split these up into more explicit options, however I don't like moving even more rootless logic to a helper script.

Having a mode whereby a user can just specify something like `--rootless` is extremely nice. I think the issue is just with the auto detection.

I guess this is fine to fix the bug, but would definitely like to see work around the UX for rootless.

**Comment by cpuguy83** [comment]
Also, wondering if we can reproduce this issue in a test.

**Comment by tonistiigi** [comment]
I don't understand the issue precisely and why we can't just ignore the `$USER` when `--rootless` is not set , but adding `DOCKER_HONOR_XDG` seems to limit the cases where rootless works without any extra configuration. Also, not sure about the `--exec-opt`. Normally we would not have options like these if we didn't have rootless mode. And in the future if we can support these with rootless they would become meaningless and a security risk.

**Comment by AkihiroSuda** [comment]
A quick workaround for the boot2docker issue https://github.com/moby/moby/issues/39009 would be to just disable setting`--rootless`  automatically depending on the `$USER` value, but it won't fix the rootful-docker-in-rootless-docker issue https://github.com/moby/moby/issues/38702 which only requires `--exec-opt native.cgroupdriver=none --exec-opt native.restrict_oom_score_adj=1`

**Comment by tianon** [comment]
I just used `env -i` to scrub extra variables while invoking
`/etc/init.d/docker` during boot to get the release out for folks to test.

Longer-term I'd love to see a proper fix like this one instead (that
doesn't rely on the inconsistent `USER` variable). 👍

Hopefully in the next Docker beta release? 🙏

**Comment by tianon** [comment]
I'd definitely love to hear more about the edge cases checking USER is
supposed to help with though, because there's got to be something more
reliable we could do. 😅

**Comment by AkihiroSuda** [comment]
@cpuguy83 @tonistiigi WDYT on https://github.com/moby/moby/pull/39024#issuecomment-481082115 ?

**Comment by AkihiroSuda** [comment]
rebased

**Comment by tonistiigi** [comment]
>  but it won't fix the rootful-docker-in-rootless-docker issue #38702 which only requires --exec-opt native.cgroupdriver=none --exec-opt native.restrict_oom_score_adj=1

When I tried this in https://github.com/moby/moby/issues/38817 I still had to set `--rootless`. I did not invoke dockerd through rootlesskit, so `--rootless` there meant "I know I'm already in rootless container so reuse parent rootless limits?". So can't we just run tests for these cases on allow them when `--rootless` has been set as `--rootless` already implies that user is ok with these limitations.

**Comment by AkihiroSuda** [comment]
The `--rootless` monolith is not extensible for future potential support of `--exec-opt native.cgroupdriver=cgroup2-rootless`  or `--exec-opt native.cgroupdriver=pam_cgfs`

File: contrib/dockerd-rootless.sh (hunk 1 above)
Comment (AkihiroSuda) [reply contrib/dockerd-rootless.sh]: merged in https://github.com/docker/docker-install/pull/109

**Comment by AkihiroSuda** [comment]
I'd like to either get this merged or find an alternative way soon, so that we can provide almost stable package before DockerCon.

**Comment by tonistiigi** [comment]
> The --rootless monolith is not extensible for future potential support of --exec-opt native.cgroupdriver=cgroup2-rootless or --exec-opt native.cgroupdriver=pam_cgfs

I think we can discuss it again when these become a reality. If there is a usecase to actually choose between different drivers, exec-opt is probably a right flag. But atm we are just determining reduced featureset so that rootless mode(dockerd inside rootless container) can start.

**Comment by AkihiroSuda** [comment]
Note that `--exec-opt native.cgroupdriver=<driver>` is not a new stuff: https://docs.docker.com/engine/reference/commandline/dockerd/#docker-runtime-execution-options

> The native.cgroupdriver option specifies the management of the container’s cgroups. You can only specify cgroupfs or systemd. If you specify systemd and it is not available, the system errors out. If you omit the native.cgroupdriver option, cgroupfs is used.

And it had been very confusing that the cgroup driver value was substantially ignored when the daemon is running in "rootless" mode. So I feel `--exec-opt native.cgroupdriver=none` is better for clarifying that cgroup stuffs are disabled.

**Comment by tonistiigi** [comment]
> Note that --exec-opt native.cgroupdriver=<driver> is not a new stuff: 

Yes, I had forgotten that. So shouldn't be an issue if there is a need for a special driver for cgroup2 in the future(although ideally it can still be avoided).

My thinking is that if user is running rootless containers `--rootless` should be enough and `--exec-opt native.cgroupdriver=none` should not be needed. And if user is running without any interest in rootless containers there shouldn't be an option for `--exec-opt native.cgroupdriver=none`. Unless there is a practical case other usecase. Using none has a security impact what users might not realize and is generally a bad idea if your system is cgroup capable.

**Comment by AkihiroSuda** [comment]
> And if user is running without any interest in rootless containers there shouldn't be an option for --exec-opt native.cgroupdriver=none. Unless there is a practical case other usecase. Using none has a security impact what users might not realize and is generally a bad idea if your system is cgroup capable.

Updated PR to disallow `native.cgroupdriver=none` when the daemon is not running in userns. (`daemon/daemon_unix.go:VerifyCgroupDriver()`)

WDYT?

> My thinking is that if user is running rootless containers --rootless should be enough

As user still need to use `dockerd-rootless.sh`, this does not seem to improve UX.
In future we may embed `dockerd-rootless.sh` equivalent to `dockerd` though.

Also, the behavior of `--rootless` is likely to be inconsistent and confusing when running `dockerd` as a userns root in rootless docker (#38702), because XDG must be ignored and `rootlesskit-docker-proxy` must not be set in #38702 usecase.

**Comment by AkihiroSuda** [comment]
It might be good to reserve `--rootless` flag for embedding `dockerd-rootless.sh` equivalent to `dockerd` in future.

**Comment by cpuguy83** [comment]
I agree with @tonistiigi, this should just be encompassed by `--rootless`, which would hopefully just get better (require less from docker-rootless.sh) over time.

**Comment by AkihiroSuda** [comment]
> this should just be encompassed by --rootless, which would hopefully just get better (require less from docker-rootless.sh) over time.

Do you mean you want `--rootless` to do equivalent of `DOCKER_HONOR_XDG=1` (implementation might be hard) and `--userland-proxy-path=$(which rootlesskit-docker-proxy)` over time?
The problem is that these flags needs to be ignored for dind case: https://github.com/moby/moby/pull/39024#issuecomment-484811513

**Comment by AkihiroSuda** [comment]
How about combined approach - retore `--rootless` as the shorthard for `DOCKER_HONOR_XDG=1 --exec-opt native.cgroupdriver=none --exec-opt native.restrict_oom_score_adj=1 --userland-proxy --userland-proxy-path=$which rootlesskit-docker-proxy)`.

For typical use case, user only needs to set `--rootless` via `dockerd-rootless.sh`.
When running dockerd as root in rootless Docker, user must not specify `--rootless` but must specify `--exec-opt native.cgroupdriver=none --exec-opt native.restrict_oom_score_adj=1`.

**Comment by cpuguy83** [comment]
Coming from a mostly naive place here... but I would say let's not set `--userland-proxy-path` with `--rootless`.

Why must the XDG dir be ignored (with --rootless) in root-in-userns case?

**Comment by AkihiroSuda** [comment]
Closing per https://github.com/moby/moby/pull/39138

> Why must the XDG dir be ignored (with --rootless) in root-in-userns case?

Because "rootful"-in-rootless dockerd is expected to behave almost same as rootful dockerd.
So, the API path should be `/var/run/docker.sock`, not `$XDG_RUNTIME_DIR/docker.sock`.

**Comment by cpuguy83** [comment]
But if the user set --rootless why would they expect it to behave the same as rootful?

**Comment by AkihiroSuda** [comment]
same as rootful but cgroup disabled

**Pull request closed without merging**